Foundational Software

mre opened the pull request.

mre committed:
draft

mre committed:
edit

mre committed:
typos

mre committed:
wording

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+
+I was at Rust Week where Niko Matsakis gave his talk "Our Vision for Rust" in which he made a profound and insightful statement:
+
+**Rust is a language for building *foundational software*.**
```
```suggestion
> Rust is a language for building *foundational software*.
```

It's a quote. Use the quoting functionality. Besides the semantic benefits, it's also _more_ of an emphasis than just bolding the text.

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+> The first contribution that I had to the Linux kernel was some fix for the ext3 file system. It was a very emotional moment for me. I sent a patch to the Linux Kernel and then I saw an email response from Al Viro - one of those developers I'd only heard about and dreamed of meeting someday.
+> He responded, **'I've never seen code this bad in my life. You managed to introduce three new bugs in two new lines of code. People like you should never be allowed to get close to a keyboard again.'**
+> That was my introduction to Linux. 
+> -- from the [Top Shelf Podcast](https://www.youtube.com/watch?v=biBLEKm2dtY&t=307s)
```
This is not an interview or clip show, it's a blog post and "here is" is definitely the wrong style for this medium.

Markdown doesn't have special syntax for `<cite>` and `<figure>` (the semantic correct way of attribution), but `> --` is wrong too.

```suggestion
> The first contribution that I had to the Linux kernel was some fix for the ext3 file system. It was a very emotional moment for me. I sent a patch to the Linux Kernel and then I saw an email response from Al Viro - one of those developers I'd only heard about and dreamed of meeting someday.
> He responded, **'I've never seen code this bad in my life. You managed to introduce three new bugs in two new lines of code. People like you should never be allowed to get close to a keyboard again.'**
> That was my introduction to Linux.

-- Glauber Costa, co-founder of Torso, on the [Top Shelf Podcast](https://www.youtube.com/watch?v=biBLEKm2dtY&t=307s)
```

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+> That was my introduction to Linux. 
+> -- from the [Top Shelf Podcast](https://www.youtube.com/watch?v=biBLEKm2dtY&t=307s)
+
+Glauber previously worked at Red Hat, Parallels, ScyllaDB, and Datadog on schedulers, databases, and performance optimizations. He knows what he's doing.
```
Did he work there before Turso or before sending the patch? Not sure this statement shows the experience at the time he send the patch.

mre replied:
rephrasing. he sent the patch before working there.

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+To me, it's about breaking down the barriers to entry and making larger parts of the software stack accessible to more people. 
+
+> We are committed to providing a friendly, safe and welcoming environment for all
+> -- [The Rust Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)
```
```suggestion
> We are committed to providing a friendly, safe and welcoming environment for all

-- [The Rust Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)
```

same note about `<cite>` as above

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+
+I found that my clients often don't use Rust as a C++ replacement.
+Many clients don't even have any C++ in production in the first place.
+They also don't need to work on the hardware-software interface or spend their time in low-level code.
```
nit: C++ isn't the default choice for hardware interaction, C is. By a large margin.

mre replied:
true, but I think it's understandable

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+These services need to be robust and reliable and serve as platforms for other code that might or might not be written in Rust.
+This is, in my opinion, the core value proposition of Rust: to build things that form the bedrock of critical infrastructure and must operate reliably for years.
+
+Rust is a day-2-language: All of the problems that you have during the lifecycle of your application surface early in development.
```
I don't quite get what you mean with "day-2-language"? Like you only start using Rust on day 2, after you spent the whole of day 1 with JavaScript or Python? Or that it only starts to shine on day 2?

Now that I have written it down it's clear to me what you meant, but "surface early in development" send me in the wrong direction. "Early" = day 1.

mre replied:
rephrased

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+This mindset takes away the focus from Rust as a C++ replacement and also explains why so many teams which use languages like Python, TypeScript, and Kotlin are attracted by Rust.
+
+What is less often talked about is that Rust is a language that enables people to move across domain boundaries: from embedded to cloud, from data science to developer tooling. 
+I don't know any other language that's so versatile. 
```
Python and JavaScript are this versatile

Python:
embedded: https://micropython.org/
cloud: https://github.com/kubernetes-client/python
data science: https://numpy.org/
developer tooling: https://github.com/httpie/cli

JavaScript:
embedded: https://www.espruino.com/
cloud: https://www.pulumi.com/docs/iac/languages-sdks/javascript/
data science: https://www.tensorflow.org/js
developer tooling: https://github.com/microsoft/vscode lol

You can't even make the argument that Rust is better at any of these than JS/Python, mainly because of the explorative nature of data science.

You could try arguing that these examples are mostly invalid because they leverage huge code sections written in C or C++ and Rust is good at these problems with 100% Rust native solutions. But you would essentially disqualify all interpreted languages which doesn't seem like a fair argument.

mre replied:
fair. rephrased.

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+
+Rust serves as a catalyst: companies start using it for critical software but then, as they get more comfortable with the language, expand into using it in other areas of their business:
+
+> I've seen it play out as we built Aurora DSQL - we chose Rust for the new dataplane components, and started off developing other components with other tools. The control plane in Kotlin, operations tools in Typescript, etc. Standard "right tool for the job" stuff. But, as the team has become more and more familiar and comfortable with Rust, it's become the way everything is built. A lot of this ...
+>
+> -- [Marc Brooker, engineer at Amazon Web Services in Seattle on lobste.rs](https://smallcultfollowing.com/babysteps/blog/2025/05/15/10-years-of-rust/)
```
```suggestion
> I've seen it play out as we built Aurora DSQL - we chose Rust for the new dataplane components, and started off developing other components with other tools. The control plane in Kotlin, operations tools in Typescript, etc. Standard "right tool for the job" stuff. But, as the team has become more and more familiar and comfortable with Rust, it's become the way everything is built. A lot of this is because we've seen the benefits of Rust, but at least some is because the team just enjoys writing Rust.

-- [Marc Brooker, engineer at Amazon Web Services in Seattle on lobste.rs](https://smallcultfollowing.com/babysteps/blog/2025/05/15/10-years-of-rust/)
```

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+They reach for loftier goals because they can.
+The fact they don't have to deal with security issues anymore enables better affordances.
+From my conversations with other Rustaceans, we all made the same observation: suddenly we can build more ambitious projects that we never dared tackling before, such as writing a CPU emulator because we can.
+For fun.
```
Bad example. You start the argument with "people don't settle for building 5 story houses anymore, once they reach the 4th floor they are already planning for the 8th" and then turn it into "people don't settle for building 5 story houses anymore, the are also constructing rally cars on the weekend". You don't build a for-fun CPU emulator because in your day job it got easier to build complex programs. People have done that in JavaScript, a notorious 1-story-house language.

mre replied:
removing example

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+and many, many more.
+
+All of these things power organizations and *must not fail* or at least do so *gracefully*.
+My clients and the companies listed on our [podcast page](/podcast) all have one thing in common: 
```
```suggestion
My clients and the companies I interviewed on our [podcast](/podcast) all have one thing in common: 
```

Makes it sound less like you interviewed your clients for the podcast and more like these are two distinct groups.

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+My clients and the companies listed on our [podcast page](/podcast) all have one thing in common: 
+They work on Rust projects that are not on the sideline, but front and center, and they shape the future of their infrastructure.
+
+Rust is useful in situations where the ["worse is better" philosophy falls apart](https://dreamsongs.com/Files/worse-is-worse.pdf): it's a language for building the "right thing":
```
```suggestion
Rust is useful in situations where the ["worse is better" philosophy falls apart](https://dreamsongs.com/Files/worse-is-worse.pdf); it's a language for building the "right thing":
```

Alternatively a comma. Fantastic link btw!

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+I think many companies will choose Rust to build their future platforms on.
+As such, it competes with C++ as much as it does with Kotlin or Python.
+
+I believe that we should shift the focus away from memory safety (which these languages also have) and instead focus on the explicitness, expressiveness, and ecosystem of Rust that is highly competitive with these languages.
```
You could also try to argue that this heavily depends on where people are coming from. If you try to market Rust as a good replacement for "systems languages" (read C), then the biggest selling feature is memory safety. If you try to convince people coming from interpreted languages than this will change. Also, for C devs who are only asking for a faster horse (a memory safe C), then the other benefits of Rust might resemble the shape of a Ford Model T.

mre replied:
True. I could be more precise here, but I don't want to distract from my main argument and I believe it's a reasonable statement to make (in general and in the given context).

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+
+## Where Rust Shines
+
+Graydon put it this way:
+
+> "I think 'infrastructure' is a more useful way of thinking about Rust's niche than arguing over the exact boundary that defines 'systems programming'."
+>
+> "This is the essence of the systems Rust is best for writing: not flashy, not attention-grabbing, often entirely unnoticed. Just the robust and reliable necessities that enable us to get our work done, to attend to other things, confident that the system will keep humming along unattended."
+> 
+> -- From [10 Years of Stable Rust: An Infrastructure Story](https://rustfoundation.org/media/10-years-of-stable-rust-an-infrastructure-story/)
```
```suggestion
> "I think 'infrastructure' is a more useful way of thinking about Rust's niche than arguing over the exact boundary that defines 'systems programming'."
>
> "This is the essence of the systems Rust is best for writing: not flashy, not attention-grabbing, often entirely unnoticed. Just the robust and reliable necessities that enable us to get our work done, to attend to other things, confident that the system will keep humming along unattended."

-- Graydon Hoare, [10 Years of Stable Rust: An Infrastructure Story](https://rustfoundation.org/media/10-years-of-stable-rust-an-infrastructure-story/)
```

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+
+Is Rust used for real?
+
+Graydon Hoare says:
+
+> "At this point, we now know the answer: yes, Rust is used a lot. It's used for real, critical projects to do actual work by some of the largest companies in our industry. We did good."
+>
+> "[Rust is] not a great hobby language but it is a fantastic professional language, precisely because of the ease of refactors and speed of development that comes with the type system and borrow checker."
+> 
+> -- [10 Years of Stable Rust: An Infrastructure Story](https://rustfoundation.org/media/10-years-of-stable-rust-an-infrastructure-story/)
```
```suggestion
> "At this point, we now know the answer: yes, Rust is used a lot. It's used for real, critical projects to do actual work by some of the largest companies in our industry. We did good."
>
> "[Rust is] not a great hobby language but it is a fantastic professional language, precisely because of the ease of refactors and speed of development that comes with the type system and borrow checker."

-- Graydon Hoare, [10 Years of Stable Rust: An Infrastructure Story](https://rustfoundation.org/media/10-years-of-stable-rust-an-infrastructure-story/)
```

M3t0r commented on content/blog/foundational-software/index.md:
```diff
@@ -0,0 +1,221 @@
+> 
+> -- [10 Years of Stable Rust: An Infrastructure Story](https://rustfoundation.org/media/10-years-of-stable-rust-an-infrastructure-story/)
+
+To build a truly industrial-strength ecosystem, we need to remember the professional software lifecycle, which is decades long.
```
```suggestion
To build a truly industrial-strength ecosystem, we need to remember the professional software lifecycle, which is hopefully decades long.
```

M3t0r reviewed (commented):
Great post with a very nice outlook in the end!

mre committed:
Mention SDK as use-case

https://www.reddit.com/r/rust/comments/1kp74t2/rust_success_story_that_killed_rust_usage_in_a/

mre committed:
another one

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
wording

mre committed:
rephrase

mre committed:
Enhance description of Rust's versatility across domains and its correctness guarantees

mre committed:
remove confusing example

mre committed:
wording

mre committed:
update date

mre committed:
Update content/blog/foundational-software/index.md

Co-authored-by: Simon Brüggen <github@m3t0r.de>

mre committed:
Merge branch 'foundational-software' of github.com:corrode/corrode.github.io into foundational-software

mre committed:
remove extra sentence

mre committed:
headlines

mre committed:
quoting

mre commented:
Thanks for the review @M3t0r!
